packages/@best/agent/src/utils/socket.io-file.ts

```diff
@@ -0,0 +1,366 @@
+            mkdirSyncRecursively(path.dirname(dir), mode); // if does not exists, create all parents recursively
+            mkdirSyncRecursively(dir, mode); // retry
+        }
+    }
+}
+
+function createDirectoryIfNotExists(dir) {
+    try {
+        fs.accessSync(dir, fs.F_OK);
+    } catch (e) {
+        // create directory if not exists
+        mkdirSyncRecursively(dir, '0755');
+    }
+}
+
+export function SocketIOFile(socket, options) {
+    if (!socket) {
+        throw new Error('SocketIOFile requires Socket.');
+    }
+
+    this.options = options || {};
+    this.maxFileSize = +options.maxFileSize || undefined;
+    this.accepts = options.accepts || [];
+    this.chunkSize = +options.chunkSize || 10240;
+    this.transmissionDelay = options.transmissionDelay || 0;
+    this.overwrite = !!options.overwrite || false;
+    this.rename = options.rename || null;
+    this.resume = !!options.resume || false;
+
+    if (!options.uploadDir) {
+        throw new Error('No upload directory specified.');
+    }
+
+    // check directory is exists
+    if (typeof options.uploadDir === 'string') {
+        createDirectoryIfNotExists(options.uploadDir);
+    } else if (typeof options.uploadDir === 'object') {
+        for (var key in options.uploadDir) {
+            createDirectoryIfNotExists(options.uploadDir[key]);
+        }
+    } else {
+        throw new Error('options.uploadDir must be string or object array.');
+    }
+
+    this.socket = socket;
+
+    socket.on('socket.io-file::reqSync', () => {
+        socket.emit('socket.io-file::recvSync', {
+            maxFileSize: this.maxFileSize,
+            accepts: this.accepts,
+            chunkSize: this.chunkSize,
+            transmissionDelay: this.transmissionDelay,
+        });
+
+        this.emit('ready');
+    });
+
+    var self = this;
+    var uploadingFiles = {};
+
+    socket.on('socket.io-file::createFile', fileInfo => {
+        var id = fileInfo.id;
+        var uploadDir = null;
+        var uploadTo = fileInfo.uploadTo || '';
+        var data = fileInfo.data || {};
+        var filename = fileInfo.name;
+        var originalFileName = fileInfo.name;
+
+        function sendError(err) {
+            socket.emit(`socket.io-file::error::${id}`, {
+                message: err.message,
+            });
+            self.emit('error', err, {
+                uploadId: id,
+                name: filename,
+                uploadTo: uploadTo,
+                data: data,
+            });
+        }
+
+        /*
+         * Do not allow relative paths such as `../../test.js`.
+         *
+         * Ref:
+         *
+         *  * https://github.com/advisories/GHSA-9h4g-27m8-qjrg
+         *  * https://github.com/advisories/GHSA-6495-8jvh-f28x
+         *  * https://cwe.mitre.org/data/definitions/22.html
+         */
+
+        if (filename.indexOf('..') !== -1) {
+            return sendError(new Error(`Path '${filename}' is not allowed.`));
+        }
```

Comment:
Maybe change the error message to be a bit more specific like `Relative paths such as '${filename}' are not allowed.`

Reply:
> Relative paths

@dbajric The reason why I did not use _"Relative paths"_ is because something such as `a/b/c` is also a relative path. But, yes, we should improve this message.

Reply:
@dbajric I've changed the message.